### PR TITLE
Actor futures for as_completed

### LINKED ISSUE #5077: Make actor futures play with as_completed
Based on [this slack thread](https://app.slack.com/client/T019Q6SMM5Z/C019Q0QSD4J) , it would be nice if `as_completed` respected actor futures as well as normal ones. Now actor futures are not actually awaitable:
```
    def __await__(self):
        return self.result()
```
(which is a blocking call, that waits on a queue.Queue, `self.q.get()`) but it could be, since we can check for the appearance of an item in the queue (`q.empty()`) or that the result is already cached.

### PR DESCRIPTION
**Pull request by martindurant**
- [x] Closes #5077
- [x] Tests added / passed
- [x] Passes `black distributed` / `flake8 distributed` / `isort distributed`

Fixes the OP example, coded as below. This leaves the following edges:
- ActorFutures don't work with `client.gather()`, but this would be a very nice
- if the future results are not called, some cleanup at exit causes `OSError("Unable to contact Actor's worker")`

```python
import os
import platform
import random
import time
from distributed import Client, get_worker, as_completed


class Actor:
    def __init__(self, actor_id):
        self.actor_id = actor_id
        self.pid = os.getpid()
        self.hostname = platform.node()
        self.worker = get_worker().address

    def ping(self, msg=None):
        print(f"{self.actor_id} {self.pid} {self.hostname} {self.worker} {time.time()} - ping {msg or ''}")
        time.sleep(random.randint(1, 3))
        return self.actor_id


def main(client):
    anames = ["actor1", "actor2", "actor3", "actor4", "actor5"]
    afuts = [client.submit(Actor, a, actor=True) for a in anames]
    actors = {name: a for name, a in zip(anames, client.gather(afuts))}
    futs = [a.ping() for a in actors.values()]
    checks = []
    for res in as_completed(futs):
        pong = res.result()
        print(f"pong - {pong}")
        check = actors[pong].ping("Second")
        checks.append(check)
    for res in as_completed(checks):
        res.result()


if __name__ == "__main__":
    with Client() as client:
        main(client)
```

### CODE REVIEW

File: distributed/actor.py [hunk 1]
```diff
@@ -176,11 +178,11 @@ async def unwrap():
                 else:
                     # TODO: this mechanism is error prone
                     # we should endeavor to make dask's standard code work here
-                    q = Queue()
+                    q = Queue(loop=self._io_loop.asyncio_loop)
```
Comment (mrocklin) [inline distributed/actor.py:181]: I recommend using `asyncio.Queue` and `import asyncio` above.  This is slightly more verbose but also keeps things explicit for future readers.

Also, we can probably drop the `loop=` keyword.  We're always in the same event loop thread.

File: distributed/actor.py [hunk 2]
```diff
@@ -158,9 +158,11 @@ async def run_actor_function_on_worker():
                             kwargs={k: to_serialize(v) for k, v in kwargs.items()},
                         )
                     except OSError:
-                        if self._future:
+                        if self._future and not self._future.done():
                             await self._future
+                            return await run_actor_function_on_worker()
                         else:
+                            print("#########", key)
```
Comment (mrocklin) [inline distributed/actor.py:165]: Heads up, print statement

File: distributed/actor.py [hunk 3]
```diff
@@ -241,22 +243,32 @@ def __init__(self, q, io_loop, result=None):
...
+        if not hasattr(self, "_cached_result"):
+            out = await self.q.get()
             if out["status"] == "OK":
                 self._cached_result = out["result"]
             else:
                 self._cached_result = out["exception"]
-        return self.result()
+        if isinstance(self._cached_result, Exception):
+            self.status = "error"
+            raise self._cached_result
```
Comment (mrocklin) [inline distributed/actor.py:260]: If the return value is an exception then this fails inappropriately.  We probably want to condition on `out["status"]` instead.

**Review by mrocklin — COMMENTED** [review]
Thanks for doing this @martindurant .  Some initial comments below.

File: distributed/actor.py (hunk 1 above)
Comment (martindurant) [reply distributed/actor.py:181]: The queue is created outside of the async coroutine, so indeed it was picking up some other default loop, resulting in "future is on other loop" exception

File: distributed/actor.py (hunk 2 above)
Comment (martindurant) [reply distributed/actor.py:165]: WIP :)

File: distributed/actor.py [hunk 4]
```diff
@@ -241,22 +243,32 @@ def __init__(self, q, io_loop, result=None):
...
+        self.status = "finished"
+        return self._cached_result
+
+    def result(self, timeout=None):
```
[the remark this answers is not in the document]
Comment (martindurant) [reply distributed/actor.py:262]: Got it, just one thing needed cleaning up

File: distributed/actor.py [hunk 5]
```diff
@@ -158,8 +157,9 @@ async def run_actor_function_on_worker():
                             kwargs={k: to_serialize(v) for k, v in kwargs.items()},
                         )
                     except OSError:
-                        if self._future:
+                        if self._future and not self._future.done():
                             await self._future
+                            return await run_actor_function_on_worker()
```
Comment (mrocklin) [inline distributed/actor.py:162]: I'm curious, why this change?

Comment (martindurant) [reply distributed/actor.py:162]: We can't run something on an actor until a worker has reported that the actor has been initialised. This branch is for the case that the future representing the actor is still in flight (can't happen twice).

**Comment by martindurant** [comment]
Note on the test: is is not async, because in async, actor methods return coroutines instead of futures. `as_completed` works with futures only (Futures and ActorFutures, now). Should it accept coroutines, or should actors in an async context return ActorFutures too? Or maybe it's all OK as it is.

**Comment by martindurant** [comment]
Failures appear to be connection errors/timeouts in test_failed_workers.py ? Those are not obviously actor related.

File: distributed/actor.py (hunk 5 above)
Comment (mrocklin) [reply distributed/actor.py:162]: OK.  Thank you for the explanation

File: distributed/tests/test_actor.py [hunk 6]
```diff
@@ -626,6 +635,27 @@ def prop(self):
         await ac.prop
 
 
+def test_as_completed(client):
+    class Simple:
+        state = 0
+
+        def inc(self):
+            self.state += 1
+            return self.state
```
Comment (mrocklin) [inline distributed/tests/test_actor.py:644]: Maybe reuse `Counter` defined at the top of this module?

File: distributed/tests/test_actor.py [hunk 7]
```diff
@@ -626,6 +635,27 @@ def prop(self):
...
+
+    ac = client.submit(Simple, actor=True).result()
+    futs = [ac.inc() for _ in range(10)]
```
Comment (mrocklin) [inline distributed/tests/test_actor.py:647]: Nit: I prefer the pronoun `futures`.  This is what is commonly used throughout the codebase.  I'd prefer that we reduce the number of pronouns that we use if possible.

Comment (mrocklin) [reply distributed/tests/test_actor.py:647]: and fut -> future below

**Comment by mrocklin** [comment]
> Note on the test: is is not async, because in async, actor methods return coroutines instead of futures

Yup.  This is my fault.  I built actors quickly and didn't bother with async.  We probably should at some point though.

**Comment by mrocklin** [comment]
I made a couple of small suggestions.  In general this looks fine to me though.  Thank you for putting it together @martindurant

**Comment by martindurant** [comment]
I have code to make an ActorFuture be the return type of an actor method in async mode, but let it be awaitable. You could then use asyncio.gather instead of client.gather (or as_completed).

Should I post it here or wait for this to go in and make a new PR?

**Comment by mrocklin** [comment]
This seems really clean to me.  Thank you for doing this @martindurant .  Merging.
